network/onchain-discovery/src/lib.rs

```diff
@@ -0,0 +1,699 @@
+    discovery_set::DiscoverySetChangeEvent,
+    get_with_proof::{UpdateToLatestLedgerRequest, UpdateToLatestLedgerResponse},
+    trusted_state::{TrustedState, TrustedStateChange},
+    waypoint::Waypoint,
+    PeerId,
+};
+use network::{
+    connectivity_manager::{ConnectivityRequest, DiscoverySource},
+    error::NetworkError,
+    protocols::{network::Event, rpc::error::RpcError},
+};
+use rand::{rngs::SmallRng, FromEntropy, Rng};
+use std::{
+    collections::HashSet,
+    convert::{TryFrom, TryInto},
+    mem,
+    pin::Pin,
+    sync::Arc,
+    task::Context,
+    time::Duration,
+};
+use storage_client::StorageRead;
+use tokio::runtime::Handle;
+
+pub mod network_interface;
+pub mod types;
+
+pub struct OnchainDiscovery<TTicker> {
+    /// A bounded executor for handling inbound discovery set queries.
+    inbound_rpc_executor: BoundedExecutor,
+    /// Our node's PeerId.
+    peer_id: PeerId,
+    /// Our node's role (validator || fullnode).
+    role: RoleType,
+    /// The current trusted state, which keeps track of the latest verified
+    /// ledger version and validator set. This version can be ahead of other
+    /// components (e.g., state sync) since we only rely on syncing epoch change
+    /// LedgerInfo's and don't need to do tx replay.
+    trusted_state: TrustedState,
+    /// The sequence number of the most recent discovery set change event we've
+    /// processed.
+    latest_event_seq_num: u64,
+    /// An internal representation of the most recent discovery set.
+    latest_discovery_set: DiscoverySetInternal,
+    /// The set of peers we're connected to.
+    connected_peers: HashSet<PeerId>,
+    /// A channel to send requests to the network instance.
+    network_tx: OnchainDiscoveryNetworkSender,
+    /// A channel to recevie notifications from the network.
+    network_rx: OnchainDiscoveryNetworkEvents,
+    /// internal gRPC client to send read requests to Libra Storage.
+    // TODO(philiphayes): use the new storage DbReader interface.
+    storage_read_client: Arc<dyn StorageRead>,
+    /// Ticker to periodically query our peers for the latest discovery set.
+    // TODO(philiphayes): once we do a bunch of initial queries on startup, we
+    // can probably reduce the frequency.
+    peer_query_ticker: TTicker,
+    /// Ticker to query our own storage for the latest discovery set.
+    // TODO(philiphayes): we should only need to query our own storage on startup
+    // and after that rely on state-sync's reconfig notifications.
+    storage_query_ticker: TTicker,
+    /// The timeout for the whole rpc request when querying one of our peers for
+    /// their latest discovery set.
+    outbound_rpc_timeout: Duration,
+    /// Random number generator. Used to sample the next peer to query.
+    rng: SmallRng,
+    /// A local counter incremented on receiving an incoming message. Printing this in debugging
+    /// allows for easy debugging.
+    event_id: u32,
+}
+
+impl<TTicker> OnchainDiscovery<TTicker>
```

Comment:
I'd like to decouple this into two services, one that receives events from the blockchain and pushes the updates into network and perhaps another that does the networking gossip of discovery messages. Here's my motivation:

1) These provide two different functionalities
    1) obtaining the current discovery set from our local storage just to bootstrap and maintain connections
    2) gossiping discovery sets with other peers
2) These can be conveniently layered as the gossip could also acquire updates from 1.1
3) While there is certainly value in having 1.2, it doesn't seem like a strict requirement, I could also imagine a model where we have a better system integration between 1.2 and state sync

Reply:
It dawned on me a little more why I was so focused on the state sync model...

State sync explicitly has a downstream and an upstream. An upstream peer would never ask a downstream peer for state. I know @christinacelee is looking to move upstream peers to be a top level config, so that might benefit this work.

Also I think it is an interesting goal to try to have a self-organizing approach to upstream versus downstream. A peer should intrinsically be able to figure which peers are upstream versus downstream if we assume there's an explicit upstream and downstream network, i.e., closer or further from a validator or just by stated fact. At which point, network could ensure meaningful flow of data throughout the system.